### PR TITLE
fix: use math.Int with exponent to support more decimal places for coin-weight

### PR DESCRIPTION
**Pull request by troykessler**
This PR includes a fix which addresses coins that have 18 decimals which would have led to integer overflows. Furthermore it cointains an updated config.yml for local testing.

## Summary by CodeRabbit

- **New Features**
  - Added `dev-continue` target to the development server for continuous operation without reset.

- **Enhancements**
  - Improved precision and functionality for funding and reward calculations by updating data types and validation methods.
  - Enhanced configuration structure for better management of account details, pool configurations, and validator information.

- **Documentation**
  - Updated OpenAPI specifications for better clarity on coin decimals and funding amounts.

- **Testing**
  - Added new helper functions and test cases to ensure robustness of funding and reward calculations.

### CODE REVIEW

File: x/bundles/keeper/logic_bundles.go [hunk 1]
```diff
@@ -276,14 +277,13 @@ func (k Keeper) calculatePayouts(ctx sdk.Context, poolId uint64, totalPayout sdk
...
 	storageCostPerCoin := k.GetStorageCost(ctx, bundleProposal.StorageProviderId).MulInt64(int64(bundleProposal.DataSize)).QuoInt64(int64(totalPayout.Len()))
-
 	for _, coin := range totalPayout {
 		weight := whitelist[coin.Denom].CoinWeight
 		if weight.IsZero() {
 			continue
 		}
 
-		amount := storageCostPerCoin.Quo(weight)
+		amount := storageCostPerCoin.Mul(math.LegacyNewDec(10).Power(uint64(whitelist[coin.Denom].CoinDecimals))).Quo(weight)
```
[the remark this answers is not in the document]
Comment (shifty11) [reply x/bundles/keeper/logic_bundles.go]: I would also either break it down into multiple steps or just explain what is happening.

**Review by shifty11 — DISMISSED** [review]

**Comment by shifty11** [comment]
I just run the tests locally and when I change the denom of Acoin to "zcoin" the test `Produce a valid bundle with multiple coins funded and 10% inflation splitting` fails.
It seems like the inflation-splitting payout is slightly different depending on the order of the coins. Not sure if this is a bug. Did not have time to investigate further.

**Review by shifty11 — COMMENTED** [review]
I just run the tests locally and when I change the denom of Acoin to "zcoin" the test `Produce a valid bundle with multiple coins funded and 10% inflation splitting` fails.
It seems like the inflation-splitting payout is slightly different depending on the order of the coins. Not sure if this is a bug. Did not have time to investigate further.

File: proto/kyve/funders/v1beta1/params.proto [hunk 2]
```diff
@@ -12,15 +12,25 @@ message WhitelistCoinEntry {
   // coin_denom is the denom of a coin which is allowed to be funded, this value
   // needs to be unique
   string coin_denom = 1;
+  // coin_decimals are the decimals of the coin
+  uint32 coin_decimals = 2;
   // min_funding_amount is the minimum required amount of this denom that needs
-  // to be funded
-  uint64 min_funding_amount = 2;
+  // to be funded. this has to be an Int because else a coin with 18 decimals could
+  // only have a max value of 18
```
Comment (mbreithecker) [inline proto/kyve/funders/v1beta1/params.proto]: Spelling.
And I would just write that uint64 is not sufficient for coins with 18 decimals

File: testutil/integration/helpers.go [hunk 3]
```diff
@@ -1,10 +1,16 @@
 package integration
 
 import (
+	"cosmossdk.io/math"
 	globalTypes "github.com/KYVENetwork/chain/x/global/types"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+func (suite *KeeperTestSuite) MustNewIntFromStr(amount string) math.Int {
+	result, _ := math.NewIntFromString(amount)
```
Comment (mbreithecker) [inline testutil/integration/helpers.go]: throw panic if err is not nil

File: x/bundles/keeper/logic_bundles.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbreithecker) [reply x/bundles/keeper/logic_bundles.go]: Agree on making this multi line and adding some comments that it is important to multiply first

File: x/funders/types/funders.go [hunk 4]
```diff
@@ -1,11 +1,14 @@
...
+import (
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
 
 func (f *Funding) GetScore(whitelist map[string]WhitelistCoinEntry) (score uint64) {
 	for _, coin := range f.Amounts {
 		if entry, found := whitelist[coin.Denom]; found {
-			score += uint64(entry.CoinWeight.MulInt64(coin.Amount.Int64()).TruncateInt64())
+			score += uint64(entry.CoinWeight.MulInt(coin.Amount).Quo(math.LegacyNewDec(10).Power(uint64(entry.CoinDecimals))).TruncateInt64())
```
Comment (mbreithecker) [inline x/funders/types/funders.go:14]: Add comment that it is important to multiply first

File: x/bundles/keeper/keeper_suite_valid_bundles_test.go [hunk 5]
```diff
@@ -85,26 +86,26 @@ var _ = Describe("valid bundles", Ordered, func() {
 		s.App().FundersKeeper.SetParams(s.Ctx(), fundersTypes.NewParams([]*fundersTypes.WhitelistCoinEntry{
 			{
 				CoinDenom:                 globalTypes.Denom,
-				MinFundingAmount:          10 * i.KYVE,
-				MinFundingAmountPerBundle: uint64(amountPerBundle),
+				MinFundingAmount:          math.NewIntFromUint64(10 * i.KYVE),
+				MinFundingAmountPerBundle: math.NewInt(amountPerBundle),
 				CoinWeight:                math.LegacyNewDec(1),
```
Comment (mbreithecker) [inline x/bundles/keeper/keeper_suite_valid_bundles_test.go:90]: What about setting coin decimals?

File: x/funders/keeper/msg_server_update_params_test.go [hunk 6]
```diff
@@ -361,6 +368,51 @@ var _ = Describe("msg_server_update_params.go", Ordered, func() {
...
+		minFundingAmount, _ := math.NewIntFromString("1000000000000000000000")
+		minFundingAmountPerBundle, _ := math.NewIntFromString("100000000000000000000")
+
+		Expect(updatedParams.CoinWhitelist[0].CoinDenom).To(Equal("tkyve"))
+		Expect(updatedParams.CoinWhitelist[0].CoinDecimals).To(Equal(uint32(18)))
+		Expect(updatedParams.CoinWhitelist[0].MinFundingAmount).To(Equal(minFundingAmount))
+		Expect(updatedParams.CoinWhitelist[0].MinFundingAmountPerBundle).To(Equal(minFundingAmountPerBundle))
+		Expect(updatedParams.CoinWhitelist[0].CoinWeight.TruncateInt64()).To(Equal(int64(2)))
+	})
+
```
Comment (mbreithecker) [inline x/funders/keeper/msg_server_update_params_test.go:415]: Is there a test checking that the multiplication is working correctly and that with the new setup we are not exceeding the decimal places?

File: x/bundles/keeper/keeper_suite_valid_bundles_test.go (hunk 5 above)
Comment (troykessler) [reply x/bundles/keeper/keeper_suite_valid_bundles_test.go:90]: Setting zero decimals for all test cases except the one test case where I test with real world values. Else I have to adjust every test in this test suite

File: x/funders/keeper/msg_server_update_params_test.go (hunk 6 above)
Comment (troykessler) [reply x/funders/keeper/msg_server_update_params_test.go:415]: Yes here: https://github.com/KYVENetwork/chain/blob/7cb0734719a67dcdf88df1616ca90d4e2fb97417/x/bundles/keeper/keeper_suite_valid_bundles_test.go#L1595

**Comment by troykessler** [comment]
> I just run the tests locally and when I change the denom of Acoin to "zcoin" the test `Produce a valid bundle with multiple coins funded and 10% inflation splitting` fails. It seems like the inflation-splitting payout is slightly different depending on the order of the coins. Not sure if this is a bug. Did not have time to investigate further.

It was indeed because of the order of the coins. Now getting the coin with `.AmountOf(denom)` so the order does not matter anymore

**Review by mbreithecker — APPROVED** [review]

**Pull request merged**